src/HTTP/Response.php

```diff
@@ -0,0 +1,199 @@
+<?php
+
+declare(strict_types=1);
+/**
+ * SimplePie
+ *
+ * A PHP-Based RSS and Atom Feed Framework.
+ * Takes the hard work out of managing a complete RSS/Atom solution.
+ *
+ * Copyright (c) 2004-2022, Ryan Parman, Sam Sneddon, Ryan McCue, and contributors
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without modification, are
+ * permitted provided that the following conditions are met:
+ *
+ * 	* Redistributions of source code must retain the above copyright notice, this list of
+ * 	  conditions and the following disclaimer.
+ *
+ * 	* Redistributions in binary form must reproduce the above copyright notice, this list
+ * 	  of conditions and the following disclaimer in the documentation and/or other materials
+ * 	  provided with the distribution.
+ *
+ * 	* Neither the name of the SimplePie Team nor the names of its contributors may be used
+ * 	  to endorse or promote products derived from this software without specific prior
+ * 	  written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS
+ * OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY
+ * AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDERS
+ * AND CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR
+ * OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ * @package SimplePie
+ * @copyright 2004-2022 Ryan Parman, Sam Sneddon, Ryan McCue
+ * @author Ryan Parman
+ * @author Sam Sneddon
+ * @author Ryan McCue
+ * @link http://simplepie.org/ SimplePie
+ * @license http://www.opensource.org/licenses/bsd-license.php BSD License
+ */
+
+namespace SimplePie\HTTP;
+
+/**
+ * HTTP Response interface
+ *
+ * This interface must be interoperable with Psr\Http\Message\ResponseInterface
+ * @see https://www.php-fig.org/psr/psr-7/#33-psrhttpmessageresponseinterface
+ *
+ * @package SimplePie
+ * @subpackage HTTP
+ */
+interface Response
```

Comment:
Why not use `Psr\Http\Message\ResponseInterface`?

Reply:
We cannot implement PSR-7 ResponseInterface, because this would require the existence of the interfaces, so we would have to include them into the build. But that could lead to errors if the user hat the interfaces already defined themself or have them installed via composer.

Also the `ResponseInterface` is to big for out use case. We would have to implement the `with*` setters and also `UriInterface` and `StreamInterface`. However, we do not need these (yet). So I defined the `Response` interface as a subset of the `ResponseInterface` like we have done it with the PSR-16 SimpleCache Interfaces. And at a last point please note, that the `Response` interface is marked as internal. So we could later change it or replace it with PSR-7 `ResponseInterface`.

I would recommend to take a look at the follow up PR #777 where I add support for PSR-7 responses.

Reply:
To be honest, I would just drop the compiled library, rather than create more unnecessary bridging classes, forever complicating the API. People should just use Composer these days. And if we really need to support that use case, we can always adjust the build to wrap third-party libraries inside `if (!class_exists('...'))`.

As for the `with*` “setters”, the `Response` object is supposed to be immutable (a good thing IMO, the repeated calls of constructor are a travesty) so the methods are necessary.

Reply:
Dropping the compiled build could be made with release 2.0.0, but at the moment this file is used by some projects, see [here](https://github.com/LeedRSS/Leed/commit/9f2d924239a6e616d31cf47ff98a819652d686c0) or [here](https://github.com/ophian/styx/commit/13ff57d30f7647edcde0a630dfefde8664598f50). Sure, dropping BC will made refactore more easier, but I also see this as a sporting challenge. :blush: SimplePie is also used by projects like WordPress, so one requirement for changes is to keep BC, see https://github.com/simplepie/simplepie/issues/431#issuecomment-195352216

Reply:
> Dropping the compiled build could be made with release 2.0.0, but at the moment this file is used by some projects, see [here](https://github.com/LeedRSS/Leed/commit/9f2d924239a6e616d31cf47ff98a819652d686c0) or [here](https://github.com/ophian/styx/commit/13ff57d30f7647edcde0a630dfefde8664598f50).

I asked about it in https://github.com/simplepie/simplepie/issues/782.

> My goal is to deprecate the whole `File` class, and I try to touch it as little as possible. Same with the IRI class and some other classes that would be obsolete with PSR-18 support.

Would these bridge classes be dropped in favour of direct PSR-18 as well? It does not look like it from #777

> Sure, dropping BC will made refactore more easier, but I also see this as a sporting challenge. blush SimplePie is also used by projects like WordPress, so one requirement for changes is to keep BC, see [#431 (comment)](https://github.com/simplepie/simplepie/issues/431#issuecomment-195352216)

You cannot really have software that is not frozen without BC break, see [Hyrum’s law](https://www.hyrumslaw.com/) and [TeX versioning](https://en.wikipedia.org/wiki/TeX#TeX82). At best, you can select a surface that you will aim to remain backwards compatible.

For example, let’s say I have a helper function `private function get_headers(Response $response): array` in my `File` subclass. Updating to this PR will make my app crash with `Fatal error: Access level to MyFileClass::get_headers() must be public (as in class File) `.

Or to bring another example, the recent minimum PHP version bump.

Each extra class and line of code has maintenance costs so I would prefer if we did not introduce new ones unless the benefits outweigh the costs.

And I want to argue availability of a bundle should not be part of the frozen surface, which would allow us less indirections here.

Reply:
> Would these bridge classes be dropped in favour of direct PSR-18 as well? It does not look like it from #777

No. Dropping classes is a breaking change. Direct use of PSR-18 (and dropping the bridge classes) can be happen earliest with release 2.0.0.

> You cannot really have software that is not frozen without BC break, see [Hyrum’s law](https://www.hyrumslaw.com/) and [TeX versioning](https://en.wikipedia.org/wiki/TeX#TeX82). At best, you can select a surface that you will aim to remain backwards compatible.

Yes, and for the selection of this surface I recommend the Symfony BC promise: https://symfony.com/doc/current/contributing/code/bc.html
For your example with the new private helper method in an extended class there is no BC promise.

Reply:
To be clear about the BC: For me there is no problem introducing breaking changes and releasing a new major version of SimplePie every few month. But the BC wish came from @rmccue, see https://github.com/simplepie/simplepie/issues/431#issuecomment-195352216

> The main thing I ask for any changes: **Don't break backwards compatibility** - BC must be maintained above (almost) everything else. Adding more stuff is fine, refactoring is fine, as is deprecating, but don't remove anything.

While working on the refactoring I'm trying to serve this wish as best as I can. I think even after release of SimplePie 2 there will be need for backports in the latest `one-dot-x` branch for some time,